internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -135,154 +130,83 @@ func (b *batcher) genBucketID(node *commonpb.Node, resource *resourcepb.Resource
 	return fmt.Sprintf("%x", h.Sum([]byte(spanFormat)))
 }
 
-func (b *batcher) getBucket(bucketID string) *nodeBatcher {
+func (b *batcher) getBucket(bucketID string) *nodeBatch {
 	bucket, ok := b.buckets.Load(bucketID)
 	if ok {
-		return bucket.(*nodeBatcher)
+		return bucket.(*nodeBatch)
 	}
 	return nil
 }
 
 func (b *batcher) getOrAddBucket(
 	bucketID string, node *commonpb.Node, resource *resourcepb.Resource, spanFormat string,
-) *nodeBatcher {
+) *nodeBatch {
 	bucket, alreadyStored := b.buckets.LoadOrStore(
 		bucketID,
-		newNodeBucket(b, node, resource, spanFormat, b.sendBatchSize, b.timeout, b.logger),
+		newNodeBatch(b, spanFormat, node, resource),
```

Comment:
I would optimize this for the "common" case, which should be the fact that we find the batch:

```go
bucket, loaded := b.buckets.Load(bucketID)
if !loaded {
  // slow path
  bucket, loaded = b.buckets.LoadOrStore(bucketID, newNodeBatch(b, spanFormat, node, resource))
  if !loaded {
     ...
  }
}
return bucket.(*nodeBatch)
```

Up to you, but currently we always call newNodeBatch which seems inefficient.

Reply:
agreed that what you mentioned is better

Reply:
I saw the same code in tailbase probably fix that code in a different PR.

Reply:
yes, in the beginning of tail sampling there was a type with a similar LoadOrStore that took a factory function to be called when it was being stored, when that changed to sync.Map the behavior was missed (changed to sync.Map since the perf was better).